### ISSUE TITLE
[Maintenance] Generate Visual Studio and Xamarin Studio templates from a common source

### ISSUE THREAD

**Issue by hach-que** [MonoGame/MonoGame#4288]
Right now MonoGame has the following project templates for various different IDEs:
- ProjectTemplates\VisualStudio2010\Android
- ProjectTemplates\VisualStudio2010\Linux
- ProjectTemplates\VisualStudio2010\OUYA
- ProjectTemplates\VisualStudio2010\Windows
- ProjectTemplates\VisualStudio2010\WindowsGL
- ProjectTemplates\VisualStudio2012\WindowsPhone
- ProjectTemplates\VisualStudio2012\WindowsStore
- ProjectTemplates\VisualStudio2012\WindowsStoreXaml
- ProjectTemplates\VisualStudio2013\WindowsPhone
- ProjectTemplates\VisualStudio2013\WindowsPhone8.1
- ProjectTemplates\VisualStudio2013\WindowsStore
- ProjectTemplates\VisualStudio2013\WindowsStoreXaml
- ProjectTemplates\VisualStudio2013\WindowsUniversal
- ProjectTemplates\VisualStudio2013\WindowsUAP
- IDE\MonoDevelop\ (and all the various template files under here)

We can probably make it easier to maintain all of these variants if we generate them - this would have the added bonus that if we want to quickly test the current version of MonoGame that we're working on as an actual app, we can just create a new project from those templates rather than having to actually install the templates into Visual Studio / Xamarin Studio first (especially since all this really is is just fancy text replacement).

**Comment by cra0zy** [comment]
I had a similar idea about this with #3954 a while back.

**Comment by tomspilman** [comment]
Yeah...  I think ideally we want what @cra0zy suggested in #3954.

The template generation systems in VS and XS both stink in different ways.  Neither is doing us any favors or solve any of the hard issues.

So IMO the idea of sidestepping them entirely and have it just launch our own project creation is fine.  The key still is still providing some basic hook so that you can launch our project creation from VS/XS in a way familiar to most users.

**Comment by hach-que** [comment]
This task was far more focused on just unifying the VS / XS templates into some common source and generating the real templates with a shell script and some text replacements so that there isn't effectively two slightly different copies of the same thing in the repo. (i.e. not user facing at all)

There is plans for a more user facing project creation wizard tool that utilizes Protobuild to do new project generation (some sort of GUI that kicks off when running it inside a folder without a module). That seems like it has a fair amount of overlap with #3954, but I didn't want to bring it up for obvious reasons.

**Comment by hach-que** [comment]
So I've done some work on the project creation tool I was talking about; it doesn't do the project creation yet, but I'm getting the basis down for having it work on all platforms (Windows, Linux, Mac) natively.

A few important points:

  * **It can be branded:** You ship or embed an XML file that switches out all of the Protobuild logos and names for MonoGame, so your users don't see that stuff at all.
  * **It's a seperate tool to Protobuild.exe:** This isn't embedded or referenced by the Protobuild project at all; it's an entirely separate tool (that just happens to use it to generate the projects during creation).
  * **It's a single exe:** So you can create the branded version and offer it as a download link, or put it in the installation folder alongside MonoGame and have VS / XS launch it (or add a Start Menu link to it).  Again, everywhere here it would have the MonoGame logo and name.

Here's the project link (what I've done so far): https://github.com/hach-que/Protobuild.Manager

And here's a screenshot of the opening screen (just picture this with MonoGame icon / name / etc. instead):

![Opening screen](http://i.imgur.com/IlsCmEf.png)

**Comment by cra0zy** [comment]
Just few things I want to say:
 - wow it looks nice
 - my original idea was more of a tool that looks and works like: https://github.com/libgdx/libgdx/wiki/Project-Setup-Gradle I don't see a point in adding recent project/open project functions since MonoDevelop/Visual Studio already have those
 - one more thing about my original idea, which you can use very easily, is to split the tool into a dll and an exe. Exe would function just as a window frame, while the dll would contain the widget and control that could be embedded inside Xamarin Studio/Visual Studio project creation wizard.
    - using mono xwt really makes this part easy

**Comment by hach-que** [comment]
> my original idea was more of a tool that looks and works like: https://github.com/libgdx/libgdx/wiki/Project-Setup-Gradle I don't see a point in adding recent project/open project functions since MonoDevelop/Visual Studio already have those

For non-MonoGame stuff, I intend for this to be like "open the Protobuild module for your game and then you can switch platforms".  You can see a video of this working here: https://www.youtube.com/watch?v=v7j505P5AkM  Obviously this is significantly easier than instructing people to run Protobuild.

> one more thing about my original idea, which you can use very easily, is to split the tool into a dll and an exe. Exe would function just as a window frame, while the dll would contain the widget and control that could be embedded inside Xamarin Studio/Visual Studio project creation wizard.

My plan here was just to have a VS project template like those "Install Some Extra Functionality" ones they provide, but instead it'd be "New MonoGame Project" and that would launch this tool (where you'd then pick the actual template to use + the location to create the project).

> using mono xwt really makes this part easy

I'm not using Xwt because I want to ship a single executable.  The only non-standard dependency is webkit-sharp on Linux, but that's because it's not installed with the `mono-complete` package.  If it's missing, users are instructed to install it through their package manager.

Actually because it's a single executable, it can in some ways replace the installers.  You can make it so the first time you run this tool, it installs the VS / XS templates to hook things up, as well as place the tool in the Start Menu with some name like "Create or Open MonoGame Project".  Then you can just have on the MonoGame website like "Download Project Creation Tool for Windows / Linux / Mac" which just downloads this single exe.

**Comment by cra0zy** [comment]
When I said "using mono xwt really makes this part easy" I meant it as it would be good to use it in my case with the split dll and exe because it would allow you to get a different widget depending on what is specified as starting toolkit, for you it wouldn't be useful because you are just running an exe from the project wizard.

**Comment by hach-que** [comment]
Oh, right of course.

**Comment by tomspilman** [comment]
> You can see a video of this working here

What am I seeing there?  Why is the VS project view updating automatically like that?

**Comment by hach-que** [comment]
That is the functionality that re-generates projects for new platforms.  It only applies if your template is a Protobuild-based template (and not just like a "here's a template for one-off creation" which I presume is what you want to ship).  This functionality will be able to be disabled via the `Branding.xml` file.

It is using the Visual Studio DTE system to tell Visual Studio to save all files and open a new solution, based on the one that has an existing solution open for that project (for any given platform).

The video is a bit older, but roughly it does this:

  * Find an existing instance of Visual Studio for the project that the user wants to switch platforms on, or an existing instance of Visual Studio that has no solution open
  * If an existing instance exists:
    * Tell it to save all of the files in the solution
    * Tell it to close the solution
    * Run Protobuild.exe --sync OldPlatform
  * Run Protobuild.exe --generate NewPlatform
  * If an existing instance didn't exist:
    * Start Visual Studio with the new solution
  * If an existing instance did exist:
    * Tell it to open the new solution file that we just generated

This is the Visual Studio IDE control implementation: https://github.com/hach-que/Protobuild.Manager/blob/master/Protobuild.Manager/IDEControl/VisualStudioIDEControl.cs.  I'll be implementing something similar for Xamarin Studio / MonoDevelop on other platforms.

**Comment by tomspilman** [comment]
> It is using the Visual Studio DTE system to tell Visual Studio to save all files and open a new solution

That is neat...  but getting back into the realm of things we do not need.  That has to be something we can turn off in the end.

**Comment by hach-que** [comment]
Yeah, I intend for every aspect of this to be brandable / customizable.

So there's a `Branding.xml` file here which has the default settings: https://github.com/hach-que/Protobuild.Manager/blob/master/Protobuild.Manager/Branding.xml.  This is an embedded resource inside the app.

There are two ways to brand it, you can either modify the embedded resources if you want to ship a single application, or you can just put a new `Branding.xml` file next to the tool and it will use that one instead.  The latter option is mainly for use cases with the installers, where it's easier to just unpack a bunch of files and add a Start Menu link.

**Comment by cra0zy** [comment]
Just wireframes: https://youtu.be/ZEGUW6R18No

**Comment by cra0zy** [comment]
GUI and GUI Logic done: https://youtu.be/IvwRahT1HTY

**Comment by hach-que** [comment]
@cra0zy If you are going to implement that, please make sure you include Protobuild as a project type.

**Comment by cra0zy** [comment]
> @cra0zy If you are going to implement that, please make sure you include Protobuild as a project type.

Added

**Comment by hach-que** [comment]
Also if you use the following template values:

`{PROJECT_NAME}` the name as typed in by the user.
`{PROJECT_SAFE_NAME}` the name as a valid C# identifier.
`{PROJECT_XML_NAME}` the name as typed in by the user, escaped with XML encoding.
`{PROJECT_SAFE_XML_NAME}` the name as a valid C# identifier, escaped with XML encoding.

then you can use Protobuild to generate all of those templates for all of the sample solutions.  Any file with those text strings in them will have them substituted when instantiating the template, and any file with those text strings in the name will be renamed on copy when instantiating the template.

**Comment by hach-que** [comment]
(At this point we're both trying to build the same thing, so I'm not sure what the end of goal of duplicating our work is here?)

**Comment by cra0zy** [comment]
> (At this point we're both trying to build the same thing, so I'm not sure what the end of goal of duplicating our work is here?)

Similar, yes, same, no.

In simple words, your tool is meant to be used with Protobuild, mine is meant to be used with MonoGame.

**Comment by hach-que** [comment]
Mine is also meant to be used with MonoGame (as explained above).  Primarily the reason it's being built the way it is is so that it can be used by MonoGame upstream.

**Comment by tomspilman** [comment]
I see it as two different paths:

  1. A cross-platform tool that mimics VS's template system and with more flexibility.
  2. A cross-platform tool that generates a Protobuild project.

1 is what we asked for and need.  2 could work as well but I am concerned that Protobuild's priorities will conflict with MonoGame priorities.

One thing I think @cra0zy seems to be addressing that I don't see in Protobuild is generating projects for multiple platforms at once.  Protobuild seems to be taking some "switch project types" approach that isn't what we are looking for.

**Comment by hach-que** [comment]
The solution I'm working on does generate multiple solutions at once; it does that for project templates that aren't using Protobuild (where you obviously can't switch out later).

**Comment by hach-que** [comment]
Project creation actually working: https://youtu.be/uxnvCNX95ts

This still has some rough edges (like it doesn't yet pull the template from a local source, it goes online just because that's the only working template I have at the moment).  I will address these issues soon.

**Comment by hach-que** [comment]
(for some reason that video cut off right before the demo of playing the game, but the template does work out-of-the-box)

**Comment by tomspilman** [comment]
> Project creation actually working

Seems like it depends on git clones to get the template... is that required?  We plan to ship the templates with the rest of the SDK.

Also it seemed that the generated project includes MonoGame projects.... is that required?  We would simply have the project linking against the shipped binaries and/or nuget.

**Comment by hach-que** [comment]
No, that's just a consequence of me using the existing template I have in Git.  I 100% intend to pull templates from disk (that's what I'm working on now).

The user will also be able to pick what type of reference to MonoGame they want to make (Git submodule, NuGet, etc.)

**Comment by cra0zy** [comment]
Yey, a build off, this is gonna be fun :D

**Comment by hach-que** [comment]
Actually I hate racing implementations; it tends to lead to multiple worse solutions rather than one good one because the only thing that matters is being first :/

**Comment by cra0zy** [comment]
> Actually I hate racing implementations; it tends to lead to multiple worse solutions rather than one good one because the only thing that matters is being first :/

I didn't mean it in a racing context, even if one of us finishes their PR probably isn't gonna get merged until the other person finishes as well.

**Comment by hach-que** [comment]
Quick update on what the project creation form is looking like now:

![image](https://cloud.githubusercontent.com/assets/504826/11617383/3b2ab60a-9ce2-11e5-8190-01164e4ca9a0.png)

![image](https://cloud.githubusercontent.com/assets/504826/11617386/4a53b276-9ce2-11e5-8166-14d563b6caeb.png)

![image](https://cloud.githubusercontent.com/assets/504826/11617389/57deab08-9ce2-11e5-8c43-68e980ba861f.png)

The additional options makes use of an overlay system, so you can add additional options for different templates.  The selected overlay will be copied on top of the template; this allows you to have an overlay for referencing NuGet packages, and only specify the difference in file structure from the base template.  Then you can take this single NuGet overlay and use it for all of the templates you offer...

**Comment by KonajuGames** [comment]
Is this a Windows only UI?

**Comment by hach-que** [comment]
No, the Protobuild Manager is cross-platform.  It's based off a cross-platform game launcher I wrote a while back, which shipped as a single executable.

I haven't updated the platform-specific code in the Protobuild Manager for Linux / Mac yet, but this is just because I haven't gone into the coworking space this week where my Linux desktop + Mac is.

**Comment by hach-que** [comment]
Here's a video of the MonoGame-branded version from a few hours ago, just showing the options select: https://youtu.be/bf4qq0t3z6A.  This is what you'll get if you ran the binary executable in the `ProjectCreationTool` folder in https://github.com/mono/MonoGame/pull/4295.

**Comment by cra0zy** [comment]
That really reminds me of Unity now.

Can Protobuild generate F# solutions?

**Comment by hach-que** [comment]
There's a branch here with experimental F# support: https://github.com/hach-que/Protobuild/tree/fsharp.  No-one really expressed an interest in using F#, so I never merged it in, but we should be able to rebase and test it if there's interest.

**Comment by cra0zy** [comment]
Do you want to join forces?

My wizard is good for integration with MonoDevelop / Visual Studio, while your wizard is good for actually being in start menu and as a standalone application. With these 2 GUIs combined we could RULE THE WORLD... I mean, I mean create a good user experience.

**Comment by hach-que** [comment]
Yeah that sounds good.

My plan for Visual Studio / Xamarin Studio integration was just to launch the standalone application like VS does for "Install Tools" entries in the new project creation wizard.  Did you have any other plans for doing this?

**Comment by cra0zy** [comment]
> Did you have any other plans for doing this?

I have a Xwt Widget Wizard dll file.

I can get Gtk 2 widget so it can be embedded within MonoDevelop's creation wizard.

As for VS I am not sure how it handles stuff, I was thinking on either using WPF Widget(if it needs embedding) or just creating the window with it.

At the end of both path I would just have to call the Protobuild Generation tool from command line with specific parameters  to actually generate the new projects.

**Comment by cra0zy** [comment]
Side note Xwt addin for MonoDevelop offers a very nice project wizard (can't post any more screenshsts now, it got broken because of the latest MD update changing project template locations):
![ref](https://cloud.githubusercontent.com/assets/6773302/10087570/ec50c9fe-6314-11e5-943c-fe24217ad7c2.png)

https://github.com/sevoku/MonoDevelop.Xwt

**Comment by hach-que** [comment]
To be honest, I'm not sure how much benefit there is to embedding the tool inside MonoDevelop - at the end of the day, it's still a separate window on-screen.  That said, on Linux this tool uses GTK#, so we can embed it inside MonoDevelop just like we would an Xwt-based tool.

One thing we do need to figure out is how to control Xamarin Studio so that it opens and closes solutions as we need (I like to open the solution that's just been created).  On Windows we do this in Visual Studio via COM, but I think we'll need an add-in to do the same in Xamarin Studio.

**Comment by cra0zy** [comment]
> One thing we do need to figure out is how to control Xamarin Studio so that it opens and closes solutions as we need (I like to open the solution that's just been created).

While using Unity3D with Wine I had to write an addin that would allow me reloading of solution manually from context menu... I think I still have the code somewhere on my hard drive so leave that part to me.

**Comment by hach-que** [comment]
Also if we want to embed this process inside MonoDevelop, we can use GtkPlug / GtkSocket.  Basically it involves creating a GtkSocket inside MonoDevelop, which gives us an integer.  We then pass that integer to the new process, which creates a GtkPlug with it.  So making the GtkSocket and then running `Protobuild.Manager.exe --intent XamarinNewProject --gtk-socket-id 123` (where `123` is the number given by GTK#).  The `--intent XamarinNewProject` will tell the application to skip over the recent projects list and make the workflow entirely based around creating a new project.

**Comment by hach-que** [comment]
Actually an alternative idea is to wire up the MonoDevelop / Visual Studio side of things to these two projects:

https://github.com/hach-que/Protobuild.IDE.MonoDevelop
https://github.com/hach-que/Protobuild.IDE.VisualStudio

The MonoDevelop one is much farther along, but some of the features are waiting on upstream changes to get merged into MonoDevelop in time for MD 6.  One option here is just to #ifdef all of the newer features that are only available for Roslyn, and keep basic project creation integration around for MD 5 (which doesn't rely on newer APIs).

The Visual Studio one is barely started, but really all it needs is the project creation anyway (the more advanced features that the MonoDevelop one has won't be available in VS simply because the IDE doesn't provide the APIs it needs).

**Comment by cra0zy** [comment]
Protobuild isn't generating the solution for ProtobuildManager correctly on Linux:

```
Starting resolution of packages for Linux...
Package resolution complete.
Synchronising projects for Linux
Synchronisation complete.
Starting generation of projects for Linux
Loading: Protobuild.Manager.ThirdParty
Loading: Protobuild.Manager
Generating: (solution)
Saving service specification
Generation complete.
```

**Comment by hach-que** [comment]
Yeah I temporarily set the Platforms attribute on Protobuild.Manager to "Windows" so that the build would pass for #4295.  You'll need to modify the definition so that attribute is "Windows,Linux" instead.

**Comment by hach-que** [comment]
Here's another video of almost everything working: https://www.youtube.com/watch?v=-jEqdZHXeII

Importantly the "open project" screen that lets you switch platforms now also works for C# full projects (not just Protobuild projects).

**Comment by cra0zy** [comment]
Can Protobuild specify the version of references?

Would it be ok if I ported the tool to Gtk 3?

**Comment by hach-que** [comment]
It needs to work with whatever Mono provides out-of-the-box; we can't depend on external DLLs.  We also depend on Webkit# on Linux to provide the embedded browser, so that also needs to be compatible.

I believe with GtkPlug/GtkSocket it shouldn't matter if you have separate versions.  If you wanted to build it directly into https://github.com/hach-que/Protobuild.IDE.MonoDevelop (for modifying the MonoDevelop workflow directly), then we should put a bunch of the Protobuild Manager code inside an IncludeProject, which you can then reference in the add-in DLL so we can share the logic between the standalone application and the IDE integrations, without having external DLLs for the standalone application.

**Comment by hach-que** [comment]
The new version I just pushed now applies the overlays, so if you select normal C# projects and reference MonoGame as a NuGet package, it adds all the correct `packages.config` files.  It doesn't yet add the references themselves, but this is because `IncludeProject` doesn't support including references (only files).  So once I get that functionality added into Protobuild, I can update the templates here and we should have the project generator working for all the template types except Shared Code / PCLs (for the moment).

**Comment by tomspilman** [comment]
>  can update the templates here 

So what do the templates look like?  What is the syntax used for them?

**Comment by hach-que** [comment]
The syntax for the templates is the standard Protobuild template format. A
template is a normal Protobuild module / project, except that upon
instantiation from the template:

* All instances of {PROJECT_NAME} get replaced with the project name. This
applies both in the naming of files (so a file with that in the name will
have the actual project name substituted in), and to the contents of files.
* {PROJECT_XML_NAME} is replaced with the XML-encoded project name.
* {PROJECT_SAFE_NAME} is replaced with the project name converted to a safe
C# identifier. This means that if a project is named "123abc", you can have
the valid string "abc" used in C# files for namespace etc.
* {PROJECT_SAFE_XML_NAME} is the same as the previous one, but XML encoded.

When the user requests a non-Protobuild project, we generate for all the
platforms they request, then we delete the Build folder and Protobuild. I
suggest looking at the PR to see how the 2D platformer template is put
together.

**Comment by tomspilman** [comment]
Seems reasonable... especially if it is more reliable than the VS template system which sometimes would ignore replacements or format them weird depending on where they were put.

**Comment by tomspilman** [comment]
> then we delete the Build folder and Protobuild

Can we look to remove the Protobuild.exe inclusion?  It is getting ridiculous with a half dozen Protobuild.exe files spread thru out a project.

**Comment by hach-que** [comment]
It does remove it? There's only one Protobuild in the MonoGame repo.

**Comment by tomspilman** [comment]
> It does remove it? There's only one Protobuild in the MonoGame repo.

If we have 4 or 5 templates in the MonoGame repo.  How many copies of Protobuild.exe would be required?

**Comment by tomspilman** [comment]
FYI... currently in VS when i close the generated solution/projects it asks me to save the projects.  This is usually a sign that Protobuild is generating projects/solutions with incorrect whitespace.

**Comment by hach-que** [comment]
Zero, it uses the latest version of Protobuild, or whatever is in the
project manager's cache (which you can pre-seed by making the Windows
Installer copy MonoGame's version of Protobuild that exists in the project
root).

**Comment by hach-que** [comment]
I've noticed that happening for the Pipeline project (the save prompt), but
I never knew what was causing it. I'll look into the whitespace thing later
(actually if you file an issue that would be even better).

**Comment by tomspilman** [comment]
Great... was worried we would have a rash of Protobuild.exes littered thru the MonoGame repo.  It would only occur on generated projects if they selected the Protobuild option and not plain C# projects.

**Comment by hach-que** [comment]
Yeah, but that's expected and intentional behavior - every module in
Protobuild ships it's own version so that we have 100% backwards and
forwards compatibility across modules.

**Comment by cra0zy** [comment]
> It needs to work with whatever Mono provides out-of-the-box; we can't depend on external DLLs. We also depend on Webkit# on Linux to provide the embedded browser, so that also needs to be compatible.

Mono itself doesn't provide and gtk-sharp by default, I think.

MonoDevelop has a reference to gtk-sharp 2 by default.

Linux systems have 2 packages which you can install for gtk sharp, "gtk-sharp2" and "gtk-sharp3", I have both installed on my system.

Protobuild doesn't specify which gtk-sharp to use so MonoDevelop chooses gtk sharp 3 stuff for most of the references, except gdk sharp, don't know why.

**Comment by hach-que** [comment]
Yeah, what I mean is we can't ship additional DLLs ourselves; all the
things we reference have to be in the GAC.

**Comment by cra0zy** [comment]
Where do you want MD addin placed?

**Comment by hach-que** [comment]
Do you want to submit a PR to https://github.com/hach-que/Protobuild.IDE.MonoDevelop?  Add a new branch like `monodevelop-5` or something because `master` right now is being written against the Roslyn version of MonoDevelop which isn't out yet.  My intention is to add it as a submodule of MonoGame, ship it as a binary, or copy the code across depending on what upstream wants.

**Comment by cra0zy** [comment]
I think it's better to split the addin in 2:
 - one for Protobuild stuff
 - one for Protobuild.Manager templates

This way people can fork the second one and setup their own templates. What do you think?

**Comment by hach-que** [comment]
The Protobuild Manager doesn't need forking to drive templates, it just uses the Branding.xml file + files on disk if people want to have their own stuff; no forking required.

The add-ins should basically be able to look at registered paths to the tool and run the version appropriate for different projects.  Or to give you an example, the Visual Studio add-in is entirely driven by the templates that the tool puts in the user's template directory.  Those templates contain the path to the Protobuild Manager and the working directory (where Branding.xml is).  So when the MonoGame installer on Windows runs it will:

* Put the Protobuild.Manager.exe files somewhere inside it's install path, along with Branding.xml and all of the templates it wants
* Run `Protobuild.Manager.exe --one-time-setup` as part of it's installation.  This will install the VSIX and set up the templates to be hooked into Visual Studio which will use the MonoGame branded version.

So this shouldn't require any forking because the VSIX literally contains a class which says "launch this application and wait until the project is generated" as the wizard step, with all the input parameters coming from the template data.

I'd imagine the Xamarin Studio / MonoDevelop stuff to be similar - it looks at some directory (like ~/XamarinTemplates or something) for the templates that are generated by the tool, and those templates tell it where the location of the Protobuild Manager is to execute (and the branding and other template files).

I'm not sure what you've written into the add-in since the code isn't posted anywhere yet, so I don't know how well this aligns to what you had planned.

**Comment by hach-que** [comment]
I've published an early preview of the VSIX extension at https://visualstudiogallery.msdn.microsoft.com/fb8ac202-0059-485e-85c3-223875108407, but this is just some supporting wizard code for the templates the Protobuild Manager generates; over the weekend some time I'll make the Protobuild Manager automatically install this extension on first run (or via a command-line option).

**Comment by cra0zy** [comment]
That sounds ok.

So for MD we will need to generate templates inside "~/.local/shara/MonoDevelop/LocalInstall/Addins/ProtoBuild.Manager/" folder when the --one-time-setup is called.

As for what I have in mind, it's still to setup an embedded version (I don't mind writing more code for stuff that looks nicer), if you look at the open file in https://youtu.be/IvwRahT1HTY you'll notice that the wizard just collects the selected info in that class which can be used to launch something with arguments. I may need to setup a --no-gui oprion for Protobuild.Manager so it just works like a command line tool.

Also I was thinking on looking if there is a way to add a combobox to the MD toolbar so that the Protobuild project could have an easy platform switching system(I know how to override 2nd combobox, but it's used on android for device selection if I remember correctly).

Just a reminder, I have an exam I am getting ready for, so I don't have time to setup anything until Saturday.

**Comment by hach-que** [comment]
Yeah that all sounds good.  We'll need a --no-gui option in the future
anyway.

With regards to the MD drop-down in the UI; this is something that is
planned for MonoDevelop 6.  The MD add-in I've been working on supports
loading the Protobuild structure of projects directly into MonoDevelop,
showing the Protobuild definitions with cross-platform settings in the
Solution Explorer rather than the C# projects.  However this is dependent
on MD 6 APIs (the roslyn branch of MonoDevelop), so it can't really be used
by end-users yet.

**Comment by cra0zy** [comment]
Where can I get MD 6? I know that MD source is at: https://github.com/mono/monodevelop

**Comment by hach-que** [comment]
You just need to check out the roslyn branch in that repo. There will be a
few APIs used by my add-in that aren't upstreamed yet, I'm still working on
getting those upstreamed (I have a few open PRs).

**Comment by cra0zy** [comment]
Are you using https://github.com/hach-que/monodevelop/tree/roslyn-protobuild while developing the addin?

**Comment by hach-que** [comment]
Yes, that's my version with the extra APIs

**Comment by cra0zy** [comment]
https://youtu.be/-jEqdZHXeII?t=11s

For standard templates you should not have "Windows" and "Linux" platforms, but you should have "WindowsDX" and "DesktopGL" platforms.

**Comment by hach-que** [comment]
Neither of those are platforms (WindowsDX is just Windows).  DesktopGL is just Linux with a different name.

To be honest, it might be better to have a drop-down for users when they're making the project to select what graphics API they want.  If they pick OpenGL, then this adds `--enable MonoGame.Framework/OpenGLGraphics` to the command line so that Windows uses OpenGL.  This also allows us to offer the option of using ANGLE on desktop platforms, without an infinite permutation of platform names.

I think conventionally "DesktopGL" is a side-effect of wanting to reduce the number of platforms, because NuGet and the old project creation mechanism made it painful to maintain.  With this though, we can give game developers separate projects for platforms, and make it easy to pick different graphics APIs just through the wizard when creating the projects.  I know from my experience that in practice you really do need different platforms between Windows / Linux / Mac when doing the actual game development, because often you interact with APIs outside of MonoGame (that are specific to the platform you're actually running on).

@tomspilman @dellis1972 @KonajuGames thoughts?

**Comment by tomspilman** [comment]
> it might be better to have a drop-down for users when they're making the project to select what graphics API they want

We don't want that or will ever support that.

You start with graphics, but then someone suggest doing the same with sounds... then with songs... then with video.  Now you have a testing nightmare where everything has to be tested in all combinations.

No thanks.


> I think conventionally "DesktopGL" is a side-effect of wanting to reduce the number of platforms

Not really.... because of course we still have 3 platforms and you still have to test 3 times.

It was more for end users that preferred the concept of one MonoGame platform for targeting desktops.  Reducing their project management, release management, etc.


>  that in practice you really do need different platforms between Windows / Linux / Mac 

There are ways to deal with that, but generally people that want to go with this path depend on the libraries they use to abstract the platform with the same code.

Even then they can still create two different build configurations if they find the need.

**Comment by hach-que** [comment]
> We don't want that or will ever support that.
> 
> You start with graphics, but then someone suggest doing the same with sounds... then with songs... then with video. Now you have a testing nightmare where everything has to be tested in all combinations.
> 
> No thanks.

To be honest I never understood the point of OpenGL on Windows.

> Not really.... because of course we still have 3 platforms and we still have to test 3 times.
> 
> It was more for end users that preferred the concept of one MonoGame platform for targeting desktops. Reducing their project management, release management, etc.

Yeah but with Xamarin.Mac requiring different project types, at most you can only run Windows and Linux from the same MonoGame.Framework assembly.  And in practice you pretty much need different assemblies for these platforms to deal with Mono differences anyway (in your main game code or other third-party libraries).

If it was up to me (and it's not), I would just drop WindowsGL entirely.

In any case, if you don't want offer the OpenGL option for Windows in the project creation tool then I don't need to change anything.

**Comment by tomspilman** [comment]
> To be honest I never understood the point of OpenGL on Windows.

I agree... but some people are religious about it.

But it also helps serve a purpose when generally you can develop OpenGL code on Windows and have it work on Linux/Mac.  In fact this was really the original reason why Windows OpenGL was setup.


> Yeah but with Xamarin.Mac requiring different project types

Project types are irrelevant here.  People would be just using a MonoGame assembly.


> And in practice you pretty much need different assemblies for these platforms to deal with Mono differences anyway 

That doesn't seem to be a problem so far.


> I would just drop WindowsGL entirely.

That just won't happen... too many people want it to exist.

**Comment by hach-que** [comment]
> But it also helps serve a purpose when generally you can develop OpenGL code on Windows and have it work on Linux/Mac. In fact this was really the original reason why Windows OpenGL was setup.

Yeah but it won't run on Mac any more, that's the point. AFAIK you need to be using the MacOS DLL to run the games on Mac.  So really it only lets you unify Windows/Linux.

In addition, it doesn't *really* let you do this because the OpenGL drivers are extremely different between Windows and Linux, and assuming that your shader code will work on one platform and port to another without issue isn't great.  In the event that your shader code doesn't port, then you're back to needing platform-specific conditionals again or runtime checks to change out what shader effects you load.  Either way, it reduces the value that WindowsGL has.

> Project types are irrelevant here. People would be just using a MonoGame assembly.

By project types I mean the actual format of the C# projects that's generated, ProjectTypeGuids, etc and all that stuff.  Xamarin.Mac, XamMac and MonoMac have a different C# project format to .NET 4.5 on Windows / Linux.

> That doesn't seem to be a problem so far.

I ran into it pretty early on back when I was developing with MonoGame before I had Protobuild.  So for a project like Unearth.Threading we would have Unearth.Threading.Windows, Unearth.Threading.MacOS and Unearth.Threading.Linux.  Because we split our game up into multiple assemblies, it got out of hand fast (that's why I wrote Protobuild originally).

I'm not sure how many people write their game against WindowsGL and then successfully port it to Linux without changing any code.  My guess is that the number would be very small (for anything other than trivially simple games anyway).

> That just won't happen... too many people want it to exist.

Yeah I realise that, but whether it needs to be in the project creation wizard is a separate issue (I don't believe it needs to be if upstream wants to discourage it's use anyway).

**Comment by dellis1972** [comment]
@hach-que 

>Yeah but it won't run on Mac any more, that's the point. AFAIK you need to be using the MacOS DLL to >run the games on Mac. So really it only lets you unify Windows/Linux.

This is incorrect. I regularly use the Linux dll to write stuff on MacOS. The other week I did a game jam which used the DesktopGL (aka Linux/WindowsGL) and it quite happily worked on Mac and Windows (and probably Linux too) with the same binary. All using the same code base. "DesktopGL" has a real world applications. Yes it might be nicer to user WindowsDX on Windows boxes, but for those users wanting total cross platform capability (for Desktop) then OpenGL is the API of choice.. it always has been. Yes there might be issues with shaders, but as you said simple runtime checks can sort that out and unless you are doing custom shaders its not an issue.

I will admit that the "DesktopGL" is not compatible with Xamarin.Mac (but then neither is our Mac port at the moment if we are honest.. as we use an unsupported option) but it does work well with MonoKickstarter which people will probably be using to ship on Linux since asking people to install mono is usually a no go for historical reasons (better to just not mention it :P)

**Comment by dellis1972** [comment]
When DesktopGL is finally finished WindowsGL will die as well as Linux as they will be exactly the same at that point. In fact I think they are now.

**Comment by hach-que** [comment]
> it does work well with MonoKickstarter which people will probably be
using to ship on Linux since asking people to install mono is usually a no
go for historical reasons (better to just not mention it :P)

I haven't looked at how mono Kickstarter works, but if you're
redistributing mono in any way that doesn't allow users to upgrade it (e.g.
You ship it in Steam), then you need to obtain a commercial license from
Xamarin

> I will admit that the "DesktopGL" is not compatible with Xamarin.Mac (but
then neither is our Mac port at the moment if we are honest.. as we use an
unsupported option)

The PR I have in the works does change that, even the unsupported full
framework is still actually Xamarin.Mac and can be packaged, it just means
Xamarin won't give you commercial support for it

> When DesktopGL is finally finished WindowsGL will die as well as Linux as
they will be exactly the same at that point. In fact I think they are now.

DesktopGL might be nice for some people, but it always has to be optional.
If a developer wants separate projects for each platform then the project
creation tool needs to order that

**Comment by dellis1972** [comment]
>DesktopGL might be nice for some people, but it always has to be optional.
>If a developer wants separate projects for each platform then the project
>creation tool needs to allow that

It does always need to be a option (I wasn't saying otherwise), in the same way we don't want to force people to use DirectX on windows :) This is the hard part about this kind of project, we need to support allot of different use cases as everyone will want to work slightly differently :D

**Comment by cra0zy** [comment]
> I haven't looked at how mono Kickstarter works, but if you're
> redistributing mono in any way that doesn't allow users to upgrade it (e.g.
> You ship it in Steam), then you need to obtain a commercial license from
> Xamarin

Kickstarter is including mono libs in a folder so that people can upgrade them:
https://github.com/OutOfOrder/MonoKickstart/tree/master/precompiled/lib
https://github.com/OutOfOrder/MonoKickstart/tree/master/precompiled/lib64

And yes, that is allowed, you just must not bundle them inside something.

**Comment by cra0zy** [comment]
So... @hach-que are you gonna add "DesktopGL" as platform?

**Comment by hach-que** [comment]
I have the option in there for the Protobuild option, not for standard
projects yet.

I need some clarity from @tomspilman as to whether or not they want the
option included or not.

**Comment by hach-que** [comment]
Actually, it turns out I did add the option to use OpenGL on Windows.  When you pick Standard Projects, you just choose the "As a NuGet Package (Use OpenGL on Windows)" option and it changes the reference from WindowsDX to DesktopGL.

**Comment by hach-que** [comment]
![image](https://cloud.githubusercontent.com/assets/504826/11765783/6ca462ca-a1bb-11e5-8bae-17b639e793d3.png)

**Comment by cra0zy** [comment]
I don't like that... DesktopGL should be a platform and if the user wants separate projects with DesktopGL they should just call the project creation wizard one more time.

MD addin is working: https://youtu.be/tDac48t-Uvc

**Comment by cra0zy** [comment]
And a few finishing touches: https://youtu.be/vMIixCxcGgQ

**Comment by hach-que** [comment]
I don't think getting people to run the wizard multiple times if they want
a Linux platform is the good idea. Like fine if you want desktopgl as a
platform entry (it only really affects Windows), but if you omit Linux
completely people will make the assessment that MonoGame doesn't support
Linux, because they won't make the connection DesktopGL == Linux (I know i
wouldn't, it's not obvious that that is what that platform does in
comparison to all the other platform names which are actual platforms)

**Comment by cra0zy** [comment]
> I don't think getting people to run the wizard multiple times if they want
> a Linux platform is the good idea.

I win :P https://youtu.be/WKBhWXWML-g

> MonoGame doesn't support
> Linux, because they won't make the connection DesktopGL == Linux (I know i
> wouldn't, it's not obvious that that is what that platform does in
> comparison to all the other platform names which are actual platforms)

I'll make DesktopGL selected by default, that way people won't be able to miss it.

**Comment by hach-que** [comment]
> I win :P https://youtu.be/WKBhWXWML-g

This also does that, it's just more explicit about what platforms you want.

> I'll make DesktopGL selected by default, that way people won't be able to miss it.

I still don't think it's good UX, but it's up to @tomspilman about what upstream wants to present to users the first time they go to create a MonoGame project.

> MD addin is working: https://youtu.be/tDac48t-Uvc

I thought you were working on an MD addin that would launch the project creator executable.  If not, then I guess I'll have to implement my own addin (or just get the project creator to do it so there's no installation)?

**Comment by hach-que** [comment]
I've started work on the MonoDevelop add-in for the project creator: https://github.com/hach-que/Protobuild.IDE.MonoDevelop/tree/monodevelop-5

**Comment by cra0zy** [comment]
> I thought you were working on an MD addin that would launch the project creator executable. If not, then I guess I'll have to implement my own addin (or just get the project creator to do it so there's no installation)?

Well yesterday I had finally had some free time and I decided that I wanted to work on my original idea for the addin, so I did.

As to will I work on your addin... duno... we have conflicts of interest and we both strive for our own perfections...

**Comment by hach-que** [comment]
So the MonoGame-branded Protobuild Manager is now wired up to MonoDevelop: https://youtu.be/CfbWwo4-uH8

**Comment by cra0zy** [comment]
Just FYI, check MonoDevelop TemplateWizard overrides for "ConfigureWizard" and "GetFinalPageControls", you may be able to override "Configure your new project" wizard.

**Comment by hach-que** [comment]
I tried that, but ConfigureWizard happens before project name entry, so you don't get a project name to prepopulate the real project configuration tool (and then MD asks you again for a project you already generated and does weird stuff afterwards).

**Comment by cra0zy** [comment]
I looked into websharp with Gtk 3, turns out it would require manual compilation of it :(

**Comment by hach-que** [comment]
Yeah, just use GTK 2 because that's what Mono bundles out of the box.

**Comment by DDReaper** [comment]
Must say it's fantastic reading the write up on this proposal, very similar to the Project Wizard for Marmalade.
Would also be good to use as an option outside of the code editor (similar to the fist screenshot, ala Unity). So you can manage your list of MonoGame projects outside of the editor and then launch the editor from the manager.
Having the in editor wizard option (similar to the new ASP.NET wizard) is also a nice touch.

Look forward to having a test version to play around with.

**Comment by hach-que** [comment]
@DDReaper I should have a test version in the next few days, I just need to get VSIX extension bundled correctly into the Windows installer and then you'll be able to download the setup from build artifacts of https://github.com/mono/MonoGame/pull/4295.

**Comment by cra0zy** [comment]
@hach-que I just remembered something, #3748, so could your tool be setup to maybe load a specific webpage(and preferably to have a local copy of it like MD/XS does it)?

**Comment by cra0zy** [comment]
@hach-que One more thing, what do you think on the following idea, offer platforms:
 - Windows
 - Linux
 - Mac
 - DesktopGL

This way:
1) You will be happy because there are Windows, Linux, Mac platforms that are separate
2) I will be happy because there is DesktopGL platform
3) Linux users can't get confused since there is a Linux platform
4) WindowsGL users may get confused for a moment to realize that WindowsGL = DesktopGL

What do you think?

**Comment by DDReaper** [comment]
Could even be more confusing as there are many types of "desktop" :P (Windows, Linux, Android and MacOS)
I'd simply prefer WindowsDX and WindowsGL as we have now.
So having:
* WindowsDX
* WindowsGL
* Windows 8/8.1
* Windows Phone 8/8.1
* Windows Universal
* Linux
* Mac
* iOS
* Android
* Ouya (does anyone still build on this?)
* TVOS

**Comment by dellis1972** [comment]
This issue with WindowsGL is is doesn't reflect the cross platform nature of the "desktopgl" platform which will work on mac,Linux and windows with one binary. 

-----Original Message-----

**Comment by cra0zy** [comment]
@DDReaper Plans are to drop WindowsGL and Linux, and replace them with DesktopGL, which will work on all Desktop platforms.

Currently Linux and WindowsGL assemblies are the exact same.

Content Pipeline has already been updated and made WindowsGL and Linux obsolete:
https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/TargetPlatform.cs#L94

Templates will get updated with: #4232

Internally I also wanted to merge the code for WindowsGL and Linux, but @hach-que and @tomspilman couldn't agree on anything... #3891 if you want to read

**Comment by DDReaper** [comment]
Ok, understand the reason behind the GL naming then and it does make sense.  I think with good education we can get the masses to accept that.
Given protobuild though, leaving Linux and DesktopGL separate as deployable will be fine.  We can just put training examples together to show best practice and possible just use DesktopGL if they are identical. (or offer a way for people to bridge their own :D)

**Comment by hach-que** [comment]
@cra0zy I didn't have any opposition to making WindowsGL and Linux use the same set of code, I just didn't want `Protobuild.exe --generate Linux` to break.

**Comment by hach-que** [comment]
> @hach-que I just remembered something, #3748, so could your tool be setup to maybe load a specific webpage(and preferably to have a local copy of it like MD/XS does it)?

Originally I was going to put some sort of news feed into the launcher, but Recent Projects ended up taking all the space (and felt more correct for this kind of usage).  Right now I want to focus on getting this to meet the requirements of upstream, since I'm juggling this PR + the changes for NDA platforms + a whole bunch of other projects, so I need to get exactly the minimum done and then I can come back and revisit more features if I get more time.

Feel free to send PRs against any of the Protobuild repositories, but just check with @tomspilman first if what you're building lines up with his expectations (I'll want to see that or some indication that it's what upstream wants before I'll merge it into the Protobuild repos).

**Comment by cra0zy** [comment]
@hach-que What do you think on the platforms thing I suggested above?

**Comment by hach-que** [comment]
Yeah, I'm just going to add DesktopGL to the list of platforms when using the MonoGame templates.

**Comment by hach-que** [comment]
#4295 is now ready for merging, with the installer wiring up the project creation tool for Windows.  I want to get this merged in before I remove the old templates or wire it up for Linux or Mac.

If you want to try out the new project creation tool, you can download the installer for Windows here: http://teamcity.monogame.net/repository/download/MonoGame_DevelopWin/12951:id/Windows/MonoGameSetup.exe

**Comment by cra0zy** [comment]
Whenever you are posting links from monogame teamcity server always add ?guest=1 at the end so that it doesn't ask you to login. ie. http://teamcity.monogame.net/repository/download/MonoGame_DevelopWin/12951:id/Windows/MonoGameSetup.exe?guest=1

**Comment by cra0zy** [comment]
Hmm, after few months of time my train of thoughts is completely different. Here are my current thoughts:

 1. We shouldn't add any type of templates to the VS/XS/MD. This may sound strange, but think about it. There will be no addin/extension to maintain from either side. MD/XS updates can't brake the addin. We won't have to maintain 3 addins(counting Rider). All the end user will have to do is launch MonoGame tool from start menu and create the project that way. Most game engines/framework that require an IDE do it this way.

 2. About Protobuild.Manager, making a GUI thats customizable by a "Branding.xml" sounds like a bad idea. It will only be customizable to a certain degree, and every project will need it's own customization. If a project wants to use a GUI tool it's better them to create their own and customize it with a proper GUI framework. What should have been done with this IMO is for it to be a .dll file that is able to retrieve information on templates it can generate, for what platforms, what reference types are available, etc. This way you get the same tool minus the GUI which you leave to the end project.

In conclusion, I think both mine and @hach-que idea were bad, and could have been much simpler :\

@hach-que I am sorry if I offended you with anything in this post. I know that everyone gets a bit offended when you start picking on something they created.

@tomspilman @KonajuGames @dellis1972 @hach-que What do you think?

**Comment by tomspilman** [comment]
> 1.We shouldn't add any type of template to the VS/XS/MD.

I sort of agree...   less plugins is much less work for us maintaining things.  I would rather maintain one good stand alone template generator tool that is cross platform.

I just worry about people being lost...  opening VS, looking for a template, and not finding one.  We would be doing something different than most everyone else.

> making a GUI thats customizable by a "Branding.xml" sounds like a bad idea. It will only be customizable to a certain degree,

I think it is fine... as long as it does what we need it is custom enough for our needs.

> every project will need it's own customization. 

I don't know what you mean.  All we care about is if it works for MonoGame and the handful of templates we will maintain.  Can you give me an example of what these other projects are?

**Comment by hach-que** [comment]
I don't think we need to worry about the plugins much, they are extremely
bare bones in the Protobuild case because they just add a wizard hook and
generate the template files in the user's template folder. For Microsoft to
break this would mean breaking every template created for Visual Studio
since 2010, which they are clearly not going to do.

The point of having custom Branding.xml with a common tool is that this
isn't a piece of infrastructure that the MonoGame developers want to have
to maintain - maintaining the custom templates at the moment is enough
work, let alone custom building and maintaining a tool whose role is very
different from the MonoGame source code itself. This is the case with
Protobuild (the command line tool), it lets me do the hard work of making
cross-platform project generation work for every target platform on, every
host platform for every IDE, while freeing up the MonoGame developers to
focus on MonoGame itself. I don't see why this piece of infrastructure
(project creation) is any different. It frees up the devs to do more useful
work, and in the event that it somehow no longer aligns with their needs,
they can fork it without having to do any of the initial investment.

**Comment by dellis1972** [comment]
FYI Xamarin Studio 6 will have a Template Wizard interface which we can
probably hook into.
While the addin is a pain to maintain, I think having something in the IDE
which people can use it important as its the normal way dev's create
projects. Something similar to the ASP.net one where you just have one
entry "MonoGame Project" and they pick a choose the items.

For XS there will need to be a very small addin, but it will be tiny
probably.

**Comment by DDReaper** [comment]
Marmalade has the same pattern for generating projects, both as an external tool and as a vs template that starts their project generation tool.

Vs also does the same with asp templates these days, now there is only one and you customise what it generates.

So there are already precedents set for this so I think it is a good path and will mean a lot less work fo MG managing multiple templates for multiple editors

**Comment by cra0zy** [comment]
@hack-que About the "Branding" file, My idea was more of making a MVC or some other similar type of pattern, where all that is needed is just to connect some GUI to it, and it works.

This way it gets full GUI customizability without you ever having to implement any of it.

**Comment by hach-que** [comment]
That still means MonoGame developers writing and maintaining code they don't need to (then they have to write and maintain a cross-platform GUI). I don't see a technical reason for MonoGame to have to do this when it's much less work to just file a feature request.

The fact is that this GUI is being built and has been built because I need it anyway. I use the Protobuild Manager in my daily development now and it saves me *so much time* already. Especially because when I want to sync changes I don't have to remember to save in visual studio and then run a command-line tool, I just press "Sync Now" and then I can commit. 

I plan on extending it in future with more graphical interfaces to commonly used Protobuild actions for cross-platform projects (like updating Protobuild, managing new project references, creating new cross-platform projects inside the existing solution, etc)

**Comment by cra0zy** [comment]
> That still means MonoGame developers writing and maintaining code they don't need to (then they have to write and maintain a cross-platform GUI). I don't see a technical reason for MonoGame to have to do this when it's much less work to just file a feature request.

I was talking in general about tool, not about MonoGame, should have specified I guess.

> The fact is that this GUI is being built and has been built because I need it anyway. I use the Protobuild Manager in my daily development now and it saves me so much time already. Especially because when I want to sync changes I don't have to remember to save in visual studio and then run a command-line tool, I just press "Sync Now" and then I can commit.
> 
> I plan on extending it in future with more graphical interfaces to commonly used Protobuild actions for cross-platform projects (like updating Protobuild, managing new project references, creating new cross-platform projects inside the existing solution, etc)

OK, I mostly agree, idea no. 2, keep it exe, add IView and make default interface extend it, during init add an option for it to use an external .dll file as IView. This way it can work as a standalone in case no GUI dll is found, or it can load custom GUI.

**Comment by hach-que** [comment]
It's already based on a web view; you can initialize the tool to any view
state by passing the name of the workflow and base64-encoded dictionary as
command-line arguments. This is how all of the IDE addins start it. And if
you want to do more general command-line operations, you can just use
Protobuild directly, which is what the tool uses to initialize the projects.

There's no need to be having completely custom separate GUI interfaces and
plug-ins when the existing one is already cross-platform (and the tool is
intended to be able to run as a standalone program).

I don't see a real use-case where having a separate view model, writing a
plug-in architecture, versioning it and maintaining all of that (either on
my side or MonoGame's side) is easier than sending in a feature request.

**Comment by cra0zy** [comment]
Protobuild.Manager => IView => external dll?
 1) false => load internal IView
 2) true => load external IView
 => execute IView commands, don't care if it's internal or external

It will literally take you less than an hour to setup, and almost nothing to maintain :| I really don't see whats so bad with it...

**Comment by hach-que** [comment]
I have to maintain compatibility though?  What happens when a new version of the Protobuild Manager comes out and the interfaces are different?

The Protobuild Manager heavily uses web views to drive the interface, what happens if the web view implementation changes?  How do I handle that if someone is doing some custom Javascript assuming the HTML structure is the same, and I change the HTML layout? etc etc.

Whenever you have a public API people can consume, you have to think about things like versioning APIs and it adds maintenance costs for everyone.  I don't want to add a public API with versioning if it's not necessary.

**Comment by tomspilman** [comment]
Really for MonoGame all we care about is if Protobuild Manager supports what we need or not.  If it doesn't we will find some other solution.  Last thing we want to do is both use Protobuild Manager and have to heavily customize it with our own font end.  If we have to do that we might as well not use it.

**Comment by hach-que** [comment]
Exactly, which is why I don't want to add that level of complexity, especially since the Protobuild Manager has already been designed to handle both the MonoGame use cases and my own.  I don't see the need to add complexity with a public API when `Branding.xml` and a folder with the cross-IDE/cross-platform templates seems to be enough.

**Comment by cra0zy** [comment]
> I have to maintain compatibility though? What happens when a new version of the Protobuild Manager comes out and the interfaces are different?

With something like this it's kind of impossible to maintain compatibility.

> The Protobuild Manager heavily uses web views to drive the interface, what happens if the web view implementation changes? How do I handle that if someone is doing some custom Javascript assuming the HTML structure is the same, and I change the HTML layout? etc etc.

IView wouldn't be connected with Javascript/HTML in any way. All IView should do is be able(and maybe controller, if you keep it separate) is:
https://github.com/mono/MonoGame/blob/develop/Tools/Pipeline/Common/IController.cs
https://github.com/mono/MonoGame/blob/develop/Tools/Pipeline/Common/IView.cs

Basically call C# functions that get translated for the specified View. In the internal view case you would use html/javascript, but external view can use whatever GUI it wants.

> Really for MonoGame all we care about is if Protobuild Manager supports what we need or not. If it doesn't we will find some other solution. Last thing we want to do is both use Protobuild Manager and have to heavily customize it with our own font end. If we have to do that we might as well not use it.

Again, not asking for MonoGame about this... Just asking for myself so I can setup a native Gtk 3 version with HeaderBars for myself.. mmmhhhh... and also for something else I plan on working on.

**Comment by tomspilman** [comment]
> Again, not asking for MonoGame about this... Just asking for myself 

I see... ok.  If you don't mind then move this conversation to the Protobuild Manager issues:

 https://github.com/Protobuild/Protobuild.Manager/issues

Just so that we don't get any confusion on what is for MonoGame and what isn't.

**Comment by hach-que** [comment]
> With something like this it's kind of impossible to maintain compatibility.

We have extremely strong compatibility guarantees with Protobuild.  See https://github.com/Protobuild/Protobuild#overview.  I expect all public APIs in Protobuild and the Protobuild Manager to maintain this level of compatibility, which is why I'm always cautious of adding new public APIs, because they have to either be correct the first time, or they need to be versioned.

> Again, not asking for MonoGame about this... Just asking for myself so I can setup a native Gtk 3 version with HeaderBars for myself.. mmmhhhh... and also for something else I plan on working on.

I don't plan on ever supporting something like this.  The web view provides a cross-platform UI that is both consistent and is aesthetically pleasant on all platforms, unlike most cross-platform toolkits where one or more platforms look ugly or out-of-place.

If you file feature requests for the use cases you have, then I can add appropriate command-line APIs for doing what you want.  I suggest taking a look at https://secure.phabricator.com/book/phabcontrib/article/describing_problems/ for describing your use cases because it has some nice documentation on describing problems to which appropriate solutions can be built.

**Comment by Jjagg** [comment]
Everything uses the dotnet templating templates now :)
